Support `notion install node lts`

willsoto opened the pull request:
Closes #255

charlespierce commented on crates/notion-core/src/inventory/mod.rs:
```diff
@@ -277,6 +287,29 @@ impl FetchResolve<NodeDistro> for NodeCollection {
+            }) => hook.resolve("index.json")?,
+            _ => public_node_version_index(),
+        };
+        let version_opt = match_node_version(&url, |&NodeEntry { lts: ref v, .. }| v.clone())?;
```
`lts` is a Boolean value, so I don't think we need to use the `ref v` and `v.clone()`, instead I think this should be doable as just:
`|&NodeEntry { lts: val, .. }| v`

charlespierce commented on tests/smoke/notion_install.rs:
```diff
@@ -21,6 +21,18 @@ fn install_node() {
+
+    assert_that!(
+        p.node("--version"),
+        execs().with_status(0).with_stdout_contains("1.2.1")
```
Since the LTS version of Node will constantly change, we probably can't check for a specific version here. However, checking that both the `notion install node lts` command completed successfully and then that `node --version` completes successfully should be enough of a smoke test that the install is working.

willsoto replied:
😅 absolutely. I was putting a bad version in there so I could actually test that my changes were doing something but ended up getting a totally different error! Thanks for the suggestion on how to test this though 👍

charlespierce commented on crates/notion-core/src/inventory/serial.rs:
```diff
@@ -108,6 +108,7 @@ pub struct NodeEntry {
     #[serde(with = "option_version_serde")]
     pub npm: Option<Version>,
     pub files: Vec<String>,
+    pub lts: bool,
```
It looks like the issue with the test is here: I didn't look as closely as I should have at the `index.json` for Node. It looks like the `lts` property in the Node index is actually of type `false | string`. If it's false, that means no LTS, if it's a string value (like "Dubnium"), then it's that named LTS group (Node 10 is `Dubnium`, Node 8 is `Carbon`, etc.)

It looks like we'll need to have a custom deserializer for that field, that can convert a string to a boolean value. You can take a look at [this block](https://github.com/notion-cli/notion/blob/master/crates/notion-core/src/version/mod.rs#L152) for implementing a custom deserializer. In our case, we probably want to try to deserialize the value as a string, and if that succeeds we return `Ok(true)`, otherwise `Ok(false)`. Then we can annotate this line with `#[serde(deserialize_with = "<NAME OF FUNCTION>")]`, and it should be able to parse it the way we want.

willsoto replied:
I've gone ahead and implemented my best attempt at this, please let me know what you think!

charlespierce commented on crates/notion-core/src/version/mod.rs:
```diff
@@ -163,3 +167,17 @@ pub mod option_version_serde {
+    where
+        D: Deserializer<'de>,
+    {
+        match String::deserialize(deserializer) {
```
This looks perfect! The only comment I have is that I think this fits better with the parsing logic in `inventory/serial.rs`, as opposed to here with the version parsing. I believe it can also be a non-`pub` function in that module that is referenced directly in `deserialize_with`, instead of needing its own module.

willsoto commented:
Everything should be good now, thank you for the feedback! I moved `fn lts_version_serde` to the top of `inventory/serial.rs` near the only other free floating function I found in that file. I hope that is okay.

---

Unrelated but I noticed after I rebased the last time I got an error on this line telling me that `PackageConfigNotFound` is not found. Doesn't seem to be breaking anything, but wanted to let you know anyway.

https://github.com/notion-cli/notion/blob/d0fab179f19cad2cc8f791bbe86cc9c25c157be0/crates/notion-core/src/inventory/serial.rs#L255
